src/libcollections/dlist.rs

```diff
@@ -279,20 +282,27 @@ impl<T> DList<T> {
     /// b.push_back(3i);
     /// b.push_back(4);
     ///
-    /// a.append(b);
+    /// a.append_mut(&mut b);
     ///
     /// for e in a.iter() {
     ///     println!("{}", e); // prints 1, then 2, then 3, then 4
     /// }
+    /// println!("{}", b.len()); // prints 0
     /// ```
-    pub fn append(&mut self, mut other: DList<T>) {
+    pub fn append_mut(&mut self, other: &mut DList<T>) {
         match self.list_tail.resolve() {
-            None => *self = other,
+            None => {
+                self.length = other.length;
+                mem::swap(&mut self.list_head, &mut other.list_head);
+                self.list_tail = other.list_tail.take();
+                other.list_head = None;
+            },
             Some(tail) => {
                 // Carefully empty `other`.
                 let o_tail = other.list_tail.take();
                 let o_length = other.length;
                 match other.list_head.take() {
+                    // `None` branch should never happen.
                     None => return,
```

Comment:
Upon reflection it can reasonably happen, though? I see nothing wrong with appending an empty list?